math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
+     * <p>
+     * This is exposed to facilitate lower level access to the stream API for a matrix.
+     */
+    public static class SparseMatrixSpliterator implements Spliterator<SparseMatrixEntry> {
+        private final SparseMatrix m;
+        private int col;      // current colum, advanced on split or traversal
+        private int index;    // current element within column
+        private final int fence; // one past the last column to process
+
+        SparseMatrixSpliterator(SparseMatrix matrix, int col, int fence) {
+            this.m = matrix;
+            this.col = col;
+            this.index = m.colIndex[col];
+            this.fence = fence;
+        }
+
+        public void forEachRemaining(Consumer<? super SparseMatrixEntry> action) {
+            for (; col < fence; col++) {
+                for (; index < m.colIndex[col + 1]; index++) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                }
+            }
+        }
+
+        public boolean tryAdvance(Consumer<? super SparseMatrixEntry> action) {
+            if (col < fence) {
+                while (col < fence && index >= m.colIndex[col + 1]) {
+                    col++;
+                }
+                if (col < fence) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                    index++;
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                // cannot advance
+                return false;
+            }
+        }
+
+        public Spliterator<SparseMatrixEntry> trySplit() {
+            int lo = col; // divide range in half, assuming roughly constant size of all columns
+            int mid = ((lo + fence) >>> 1) & ~1; // force midpoint to be even
+            if (lo < mid) { // split out left half
+                col = mid; // reset this Spliterator's origin
+                return new SparseMatrixSpliterator(m, lo, mid);
+            } else {
+                // too small to split
+                return null;
+            }
+        }
+
+        public long estimateSize() {
+            return (long) (m.colIndex[fence] - m.colIndex[col]);
+        }
+
+        public int characteristics() {
+            return ORDERED | SIZED | IMMUTABLE | SUBSIZED;
+        }
+    }
+
+    /**
+     * Encapsulates important information about an entry in a matrix for use
+     * in streaming, including a string that leads back to the original cell
+     * so that in-place updates are possible.
+     */
+    public static class SparseMatrixEntry {
+        // these fields are exposed for direct access to simplify in-lining by the JVM
+        public final int row;
+        public final int col;
+        public double x;
+
+        // these are hidden due to internal dependency
+        private final int index;
```

Comment:
They should be removed when we remove `set` method.

Reply:
The set method shouldn't be removed.

Reply:
We want to keep SparseMatrix immutable. SparseMatrix is mostly used for matrix operations such as matrix-vector multiplication and matrix-matrix multiplication, which are typically used for optimization or eigen value decomposition. Again, we use SparseDataset for mutable operations before we convert it to SparseMatrix.

Reply:
Where is this documented?

Reply:
Other kinds of matrices are mutable. Why only make SparseMatrix immutable?

The general feeling across the implementations I have seen (in R, in Arrow, in Pandas) and the implementations I have worked on in the past (many) is that mutation of a SparseMatrix is only a problem for brittle data structures and only then if the pattern of sparsity is changed.

Leaving the set method on the element makes this safe.

It is also very useful for operations such as in-place squaring or taking the absolute value.